modules/ref_fb_module/src/struct_decoder_fb_impl.cpp

```diff
@@ -0,0 +1,289 @@
+#include <ref_fb_module/struct_decoder_fb_impl.h>
+#include <opendaq/input_port_factory.h>
+#include <opendaq/data_descriptor_ptr.h>
+#include <opendaq/event_packet_ptr.h>
+#include <opendaq/custom_log.h>
+#include <opendaq/event_packet_params.h>
+#include <opendaq/data_packet.h>
+#include <opendaq/data_packet_ptr.h>
+#include <opendaq/event_packet_ids.h>
+#include <opendaq/packet_factory.h>
+#include <opendaq/component_status_container_private_ptr.h>
+#include <opendaq/sample_type_traits.h>
+#include <opendaq/search_filter_factory.h>
+
+BEGIN_NAMESPACE_REF_FB_MODULE
+namespace StructDecoder
+{
+
+static const char* InputDisconnected = "Disconnected";
+static const char* InputConnected = "Connected";
+static const char* InputInvalid = "Invalid";
+
+StructDecoderFbImpl::StructDecoderFbImpl(const ContextPtr& ctx, const ComponentPtr& parent, const StringPtr& localId)
+    : FunctionBlock(CreateType(), ctx, parent, localId)
+{
+    createInputPorts();
+    initStatuses();
+}
+
+FunctionBlockTypePtr StructDecoderFbImpl::CreateType()
+{
+    return FunctionBlockType("RefFBModuleStructDecoder", "Struct decoder", "Decodes signals with struct data type and outputs signal for each struct component.");
+}
+
+void StructDecoderFbImpl::processSignalDescriptorChanged(const DataDescriptorPtr& inputDataDescriptor,
+                                                         const DataDescriptorPtr& inputDomainDataDescriptor)
+{
+    if (inputDataDescriptor.assigned())
+        this->inputDataDescriptor = inputDataDescriptor;
+    if (inputDomainDataDescriptor.assigned())
+        this->inputDomainDataDescriptor = inputDomainDataDescriptor;
+
+    configure();
+}
+
+void StructDecoderFbImpl::configure()
+{
+    if (!inputDataDescriptor.assigned() || !inputDomainDataDescriptor.assigned())
+    {
+        setInputStatus(InputInvalid);
+        return;
+    }
+
+    try
+    {
+        if (inputDataDescriptor.getDimensions().getCount() > 0)
+            throw std::runtime_error("Arrays not supported");
+
+        const auto inputSampleType = inputDataDescriptor.getSampleType();
+        if (inputSampleType != SampleType::Struct)
+            throw std::runtime_error("Invalid sample type");
+
+        signals.clear();
+
+        constexpr std::array<SampleType, 11> validFieldTypes{
+            SampleType::Int8,
+            SampleType::Int16,
+            SampleType::Int32,
+            SampleType::Int64,
+            SampleType::UInt8,
+            SampleType::UInt16,
+            SampleType::UInt32,
+            SampleType::UInt64,
+            SampleType::Float32,
+            SampleType::Float64,
+            SampleType::Struct,
+        };
+
+        const auto domainSignal = createAndAddSignal("__domain", inputDomainDataDescriptor);
+        domainSignal.asPtr<IComponentPrivate>().unlockAttributes(List<IString>("Visible"));
+        domainSignal.setVisible(false);
+        domainSignal.asPtr<IComponentPrivate>().lockAttributes(List<IString>("Visible"));
```

Comment:
I think there's an optional parameter in the createAndAddSignal for visibility.

Reply:
Ok will check that